streaming/base/dataset.py

```diff
@@ -189,37 +196,51 @@ def __init__(self,
         # Determine and distribute shuffle seed and shm prefix.
         seed_rng = np.random.default_rng(shuffle_seed)
         self.shuffle_seed = int(seed_rng.integers(1 << 60))
-        prefix_int = np.random.randint(1 << 24)
+        prefix_int = int(seed_rng.integers(1 << 24))
         self._prefix = f'{prefix_int:06x}'
 
+        # Create the shared memory-backed worker barrier, without its lock, which is unpickleable.
+        worker_barrier_filelock_path = os.path.join(os.path.sep, 'tmp', 'streaming', self._prefix,
+                                                    'barrier_filelock')
+        worker_barrier_shm_path = f'{self._prefix}_barrier'
+
+        # Should be a unique shared memory block filename per each StreamingDataset instantiation to avoid a conflict
+        # between a different StreamingDataset instance on a same machine.
+        if self.world.is_local_leader:
+            if os.path.exists(os.path.dirname(worker_barrier_filelock_path)):
+                raise ValueError(''.join([
+                    f'`shuffle_seed` value is same for train and val/test dataset. Please have a ',
+                    'different `shuffle_seed` value for every dataset instantiation'
+                ]))
+            # Creates a new shared memory block only from local rank 0
+            self._worker_barrier = SharedBarrier(worker_barrier_filelock_path,
+                                                 worker_barrier_shm_path, True)
+        else:
+            # Attaches to an existing shared memory block
+            self._worker_barrier = SharedBarrier(worker_barrier_filelock_path,
+                                                 worker_barrier_shm_path, False)
+
+        # Remove the lock to make it unpickleable
```

Comment:
"that makes it unpickleable"

Reply:
Fixed in https://github.com/mosaicml/streaming/pull/107/commits/ce88df6c54e914e82797c6ab504f4a01012a742a.